codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/codegen/StructureGenerator.java

```diff
@@ -117,134 +113,118 @@ public void renderStructure(Runnable runnable, boolean isInputStructure) {
+              .getNamespace()
+              .equals(refShape.getReferentId().getNamespace())
+          ) {
+            writer.addImportFromModule(
+              SmithyNameResolver.getGoModuleNameForSmithyNamespace(
+                refShape.getReferentId().getNamespace()
+              ),
+              namespace
+            );
+          }
+        } else {
+          if (
+            !member
+              .toShapeId()
+              .getNamespace()
+              .equals(targetShape.toShapeId().getNamespace()) &&
+            !targetShape.toShapeId().getNamespace().startsWith("smithy") &&
+            targetShape.asStructureShape().isPresent()
+          ) {
+            writer.addImportFromModule(
+              SmithyNameResolver.getGoModuleNameForSmithyNamespace(
+                targetShape.toShapeId().getNamespace()
+              ),
+              namespace
+            );
+          }
+        }
 
         writer.write("$L $P", memberName, memberSymbol);
       });
-
-    runnable.run();
-
-    // At this moment there is no support for the concept of modeled document structure types.
-    // We embed the NoSerde type to prevent usage of the generated structure shapes from being used
-    // as document types themselves, or part of broader document-type structures. This avoids making backwards
-    // incompatible changes if the document type representation changes if it is later annotated as a modeled
-    // document type. This restriction may be relaxed later by removing this constraint.
-    writer.write("");
-    writer.write("$L", ProtocolDocumentGenerator.NO_DOCUMENT_SERDE_TYPE_NAME);
-
     writer.closeBlock("}").write("");
+    validationGenerator.renderValidator(shape, isInputStructure);
   }
 
   /**
    * Renders an error structure and supporting methods.
    */
   private void renderErrorStructure() {
     Symbol structureSymbol = symbolProvider.toSymbol(shape);
-    writer.addUseImports(SmithyGoDependency.SMITHY);
     writer.addUseImports(SmithyGoDependency.FMT);
     ErrorTrait errorTrait = shape.expectTrait(ErrorTrait.class);
 
     // Write out a struct to hold the error data.
-    writer.writeShapeDocs(shape);
     writer
       .openBlock(
         "type $L struct {",
         "}",
         structureSymbol.getName(),
         () -> {
-          // The message is the only part of the standard APIError interface that isn't known ahead of time.
-          // Message is a pointer mostly for the sake of consistency.
-          writer.write("Message *string").write("");
-          writer.write("ErrorCodeOverride *string").write("");
-
+          writer.write(
+            "$LBaseException",
+            context.settings().getService().getName()
+          );
+          Set<String> memberNameSet = new HashSet<>();
+          // TODO: Revisit if message has to be strictly pointer or not (even with required trait).
+          // When any shape is required we don't add pointer in local service but AWS SDK does.
```

Comment:
Is this TODO fresh? Or have you already implicitly revisited this since you wrote this comment? It looks like you did below

Reply:
There's an inconsistency on how sdk models exception and how local service does. We do it right - add `@required` trait and decide to use pointers based on that. Sdk always create `*` (probably because they don't model this in aws models and has to account for null-ability, see the explanation below). Hence the TODO.